### PR TITLE
Fixing the Backslash Escape Mechanism in LIKE and ILIKE Operator

### PR DESCRIPTION
**Pull request by SiyaoIsHiding**
Regarding this issue: https://github.com/questdb/questdb/issues/2623#issuecomment-1413465936 by @bziobrowski 
I have enabled the default escape character '\\' in the pattern string after LIKE or ILIKE Operator. That means the first two scenarios in this open issue are fixed.

### CODE REVIEW

File: core/src/test/java/io/questdb/griffin/ExplainPlanTest.java [hunk 1]
```diff
@@ -921,67 +921,6 @@ public void testExplainWithJsonFormat4() throws Exception {
                 " explain (format json) select * from taba left join tabb on a1=b1  or a2=b2", null, null, false, true, true);
     }
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/ExplainPlanTest.java:922]: Looks like bad merge . 
Please restore these tests .

File: core/src/test/java/io/questdb/griffin/SqlParserTest.java [hunk 2]
```diff
@@ -379,30 +379,6 @@ public void testAtAsWithAlias() throws Exception {
         );
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/SqlParserTest.java:379]: Another bad merge . Please restore these tests .

File: core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java [hunk 3]
```diff
@@ -260,6 +260,60 @@ public void testLikeStringUnderscoreAtStartAndEnd() throws Exception {
         });
     }
 
+
+    // JANE
+    @Test
+    public void testLikeEscapePercengtage() throws Exception {
+        assertMemoryLeak(() -> {
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java:267]: Unnecessary comment .

File: core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java [hunk 4]
```diff
@@ -260,6 +260,60 @@ public void testLikeStringUnderscoreAtStartAndEnd() throws Exception {
         });
     }
 
+
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java:263]: Leftover ?

File: core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java [hunk 5]
```diff
@@ -260,6 +260,60 @@ public void testLikeStringUnderscoreAtStartAndEnd() throws Exception {
...
+
+            String sql = "create table myTable as (\n" +
+                    "select cast('Can you give me 30% discount?' as string) as comment from long_sequence(1)\n" +
+                    "union\n" +
+                    "select cast('May I get 30 USD off?' as string) as comment from long_sequence(1)\n" +
+                    "union\n" +
+                    "select cast('The target path is \\usr' as string) as comment from long_sequence(1))";
+            compiler.compile(sql, sqlExecutionContext);
+
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java:283]: There's no need to iterate explicitly . 
Please use one of assertQuery() methods .

File: core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java [hunk 6]
```diff
@@ -260,6 +260,60 @@ public void testLikeStringUnderscoreAtStartAndEnd() throws Exception {
...
+    public void testLikeEscapePercengtage() throws Exception {
+        assertMemoryLeak(() -> {
+            /**
+             * The string stored in the table is
+             * Can you give me 30% discount?
+             * The pattern I query is
+             * %30\%%
+             */
+
+
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java:275]: casting to string is unnecessary .
Only the first union-ed subquery needs column aliases .

File: core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java [hunk 7]
```diff
@@ -260,6 +260,60 @@ public void testLikeStringUnderscoreAtStartAndEnd() throws Exception {
...
+        assertMemoryLeak(() -> {
+
+            String sql = "create table myTable as (\n" +
+                    "select cast('Can you give me 30% discount?' as string) as comment from long_sequence(1)\n" +
+                    "union\n" +
+                    "select cast('May I get 30 USD off?' as string) as comment from long_sequence(1)\n" +
+                    "union\n" +
+                    "select cast('The target path is \\usr' as string) as comment from long_sequence(1))";
+            compiler.compile(sql, sqlExecutionContext);
+
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java:306]: Same as above .

File: core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java (hunk 6 above)
Comment (bziobrowski) [reply core/src/test/java/io/questdb/griffin/engine/functions/regex/ILikeFunctionFactoryTest.java:275]: You can simplify commands by creating table separately and inserting data with 
```sql
insert into myTable values (string1), (string2), string(3); 
```

File: core/src/test/java/io/questdb/griffin/engine/join/HashJoinTest.java [hunk 8]
```diff
@@ -61,7 +61,7 @@ public void testHashJoinDoesntAllocateMemoryPriorToCursorOpenAndAfterCursorClose
                     "TIMESTAMP(sensor_time);", sqlExecutionContext);
 
             compile("insert into weather_data_historical \n" +
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/engine/join/HashJoinTest.java:63]: Bad merge. Please revert .

File: core/src/test/java/io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactoryTest.java [hunk 9]
```diff
@@ -37,7 +37,7 @@ public class RegexpReplaceStrFunctionFactoryTest extends AbstractGriffinTest {
     public void testNullRegex() throws Exception {
         assertFailure(
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactoryTest.java:38]: Bad merge. Please revert .

File: core/src/test/java/io/questdb/griffin/InsertTest.java [hunk 10]
```diff
@@ -461,6 +461,26 @@ public void testInsertExplicitTimestampPos1() throws Exception {
         });
     }
 
+
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/InsertTest.java:464]: What is the goal of this test ?

File: core/src/test/java/io/questdb/griffin/AbstractSqlParserTest.java [hunk 11]
```diff
@@ -28,7 +28,6 @@
 import io.questdb.cairo.PartitionBy;
 import io.questdb.cairo.TableModel;
 import io.questdb.cairo.TableToken;
-import io.questdb.cairo.sql.RecordCursorFactory;
 import io.questdb.cairo.sql.RecordMetadata;
 import io.questdb.griffin.model.ExecutionModel;
 import io.questdb.griffin.model.ExpressionNode;
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/griffin/AbstractSqlParserTest.java:33]: Please revert these changes .

**Comment by SiyaoIsHiding** [comment]
Hi Bolek @bziobrowski ,

Thank you very much for your review and detailed suggestions! I am sorry that we have little experience in it and made our repo messy TAT. But I am committed to it and I will correct it according to your comments!

However, may I confirm the expected behavior of QuestDB string escape before I add the new tests? 

I use MySQL as a bench mark and summarize the MySQL backslash escape behavior in LIKE statements. But QuestDB and MySQL are different, as QuestDB has no backslash escape in INSERT strings. To avoid confusion, may you confirm it is the expected behavior that I should implement?
You can use this [notion link ](https://potent-asteroid-aab.notion.site/For-bziobrowski-4bf73ed6b6524e7788c83ea56c0649f4)for better formatting.

Again, thank you for your time!


### _

|  | MySQL Behavior | QuestDB Expected Behavior |
| --- | --- | --- |
| String in INSERT statement | ‘The path is \_ignore’ | ‘The path is \_ignore’ |
| Value stored | The path is \_ignore | The path is \_ignore |
| String after LIKE operator | ‘The path is \\_ignore’ | 'The path is \_ignore’ |
| Value searched | The path is \_ignore | The path is \_ignore |
| Outcome | Empty | Empty |
|  |  | |
| String in INSERT statement | ‘The path is \\_ignore’ |
| Value stored | The path is \_ignore |
| String after LIKE operator | ‘The path is \\\_ignore’ |
| Value searched | The path is \_ignore |
| Outcome | The path is \_ignore |
|  |  | |
| String in INSERT statement | 'The path is \\_ignore' | ‘The path is \_ignore’ |
| Value stored | The path is \_ignore | The path is \_ignore |
| String after LIKE operator | ‘The path is \\\\_ignore’ | 'The path is \\_ignore’ |
| Value searched | The path is \\_ignore | The path is \\_ignore |
| Outcome | The path is \_ignore | The path is \_ignore |

### %

|  | MySQL Behavior | QuestDB Expected Behavior |
| --- | --- | --- |
| String in INSERT statement | ‘Slash Percentage: \\%’ | ‘Slash Percentage: \%’ |
| Value stored | Slash Percentage: \% | Slash Percentage: \% |
| String after LIKE operator | ‘Slash Percentage: \\%’ | ‘Slash Percentage: \%’ |
| Value searched | Slash Percentage: \% | Slash Percentage: \% |
| Outcome | Slash Percentage: \% | Slash Percentage: \% |
|  |  | |
| String in INSERT statement | ‘Slash Percentage: \\%’ |
| Value stored | Slash Percentage: \% |
| String after LIKE operator | ‘Slash Percentage: \\\%’ |
| Value searched | Slash Percentage: \% |
| Outcome | Slash Percentage: \% |
|  |  | |
| String in INSERT statement | ‘Slash Percentage: \\%’ | ‘Slash Percentage: \%’ |
| Value stored | Slash Percentage: \% | Slash Percentage: \% |
| String after LIKE operator | ‘Slash Percentage: \\\\%’ | ‘Slash Percentage: \\%’ |
| Value searched | Slash Percentage: \\% | Slash Percentage: \\% |
| Outcome | Slash Percentage: \% | Slash Percentage: \% |

**Comment by SiyaoIsHiding** [comment]
@bziobrowski Although the pull request is closed because I am going to use another branch to make the pull request, may you still confirm the expected outcome above, so that I can work on it further?
Thank you very much!

**Comment by bziobrowski** [comment]
Hi @SiyaoIsHiding .  
Sorry for the delay . Your message flew under radar because PR is closed. 
Instead of MySQL it's better to compare with PostgreSQL that we try to be compatible with . 
So - better take a look at https://www.postgresql.org/docs/current/functions-matching.html .
While there's no special escape handling in insert statement, it doesn't really matter because : 
- _ and % are not special symbols to escape outside context of like expression
- one can always use raw values of \t, \n, \r, etc. if needed

The missing things are : 
- proper matching of escape symbol, e.g. assuming default escape character - \ - one has to use \\ to match it literally. \\ (or any even number) appearing before % or _ doesn't make it a literal . 
- escape clause to let user select something other than \

**Comment by SiyaoIsHiding** [comment]
Thank you very much for your suggestion! The expected behavior is crystal clear now. I will work on it and will hopefully create a new pull request this week. 😆
